controllers/clustertopology_controller_compute.go

```diff
@@ -140,9 +142,132 @@ func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clust
 	return class, nil
 }
 
-// getTemplate gets the object referenced in ref.
+// getCurrentState gets information about the current state of a Cluster by inspecting the state of the InfrastructureCluster,
+// the ControlPlane, and the MachineDeployments associated with the Cluster.
+func (r *ClusterTopologyReconciler) getCurrentState(ctx context.Context, cluster *clusterv1.Cluster, class clusterv1.ClusterClass) (*clusterTopologyState, error) {
+	clusterState := clusterTopologyState{
+		cluster: cluster,
+	}
+
+	// Reference to the InfrastructureCluster can be nil and is expected to be on the first reconcile.
+	// In this case the method should still be allowed to continue.
+	if cluster.Spec.InfrastructureRef != nil {
+		infra, err := r.getCurrentInfrastructureClusterState(ctx, cluster)
+		if err != nil {
+			return nil, err
+		}
+		clusterState.infrastructureCluster = infra
+	}
+
+	// Reference to the ControlPlane can be nil, and is expected to be on the first reconcile. In this case the method
+	// should still be allowed to continue.
+	if cluster.Spec.ControlPlaneRef != nil {
+		log.Print("going for it")
+		cp, err := r.getCurrentControlPlaneState(ctx, cluster, class)
+		log.Print(err)
+		if err != nil {
+			return nil, err
+		}
+		clusterState.controlPlane = cp
+	}
+
+	// A cluster may have zero or more MachineDeployments and a Cluster is expected to have zero MachineDeployments on
+	// first reconcile.
+	m, err := r.getCurrentMachineDeploymentState(ctx, cluster)
+	if err != nil {
+		return nil, err
+	}
+	clusterState.machineDeployments = m
+	return &clusterState, nil
+}
+
+// getCurrentInfrastructureClusterState looks for the state of the InfrastructureCluster. If a reference is set but not
+// found, either from an error or the object not being found, an error is thrown.
+func (r *ClusterTopologyReconciler) getCurrentInfrastructureClusterState(ctx context.Context, cluster *clusterv1.Cluster) (*unstructured.Unstructured, error) {
+	infra, err := r.getReference(ctx, cluster.Spec.InfrastructureRef)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read %s", cluster.Spec.InfrastructureRef.Kind)
```

Comment:
I think it would be great to also add the name here (like in getCurrentControlPlaneState)

Reply:
I'll fix this here, but I think it's better to not go over and back on error messages in favor of looking at #5067 .

Reply:
Yeah we should definitely not go back and forth. But I think the name is important and we should keep it consistent with the func below